### PR TITLE
[gh-action]: Generate a weekly report of activity

### PR DESCRIPTION
**Pull request by kevinslin**
**Description:** 

This PR creates a gh action that generates a weekly report on repo statistics. 
It delivers on the requirements specified in https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/24672
You can see the sample output here: https://github.com/kevinslin/opentelemetry-collector-contrib/issues/16

**Link to tracking Issue:** #24672

**Testing:** 

Manual testing in fork: https://github.com/kevinslin/opentelemetry-collector-contrib/actions
Example output: https://github.com/kevinslin/opentelemetry-collector-contrib/issues/17

**Documentation:** 

The architecture:
- we use `actions/github-script@v6` to make calls to the gh apis
- we require installing `js-yaml` in order to parse `metadata.yaml` files in order to get components. this dependency is installed during the github action run and not persisted

Some caveats about the logic:
- when this action runs, it looks back the previous 7 days and gets issues created in that time period, normalizing times to UTC
    - eg. if running this on wednesday (eg. 2023-08-25 17:35:00), it will scan issues from the previous wednesday (2023-08-18 00:00:00Z) to beginning of this wednesday (2023-08-28 0:00:00Z)
- this action writes the json payload of the report inside the issue. the payload is parsed by future reports to calculate deltas 
- the report issue has a custom label: `report` - this is used so we can properly filter previous issues when calculating deltas. the [github issues api](https://docs.github.com/en/rest/issues/issues?apiVersion=2022-11-28#list-repository-issues) only does filtering based on labels and `since` date

This action currently runs every Tuesday at 1AM UTC

### CODE REVIEW

**Comment by kevinslin** [comment]
didn't add a changelog entry since this was an internal facing change. happy to add an entry if we feel this warrants that

File: .github/workflows/scripts/generate-weekly-report.js
Comment (atoulme) [inline .github/workflows/scripts/generate-weekly-report.js:82]:
```suggestion
```

File: .github/workflows/scripts/generate-weekly-report.js
Comment (atoulme) [inline .github/workflows/scripts/generate-weekly-report.js:1]:
```suggestion
// Copyright The OpenTelemetry Authors
// SPDX-License-Identifier: Apache-2.0

const fs = require('fs');
```

**Comment by kevinslin** [comment]
updated issue format with gh style checkboxes: https://github.com/kevinslin/opentelemetry-collector-contrib/issues/21

**Review by atoulme — APPROVED** [review]

**Comment by kevinslin** [comment]
looks like the gh workflows requires an approval from a maintainer?

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:263]:
```suggestion
    repo: $REPO_NAME,
```

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:273]:
```suggestion
    repo: REPO_NAME,
```

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:262]: Why do use `owner: REPO_OWNER` in some places and `owner: context.payload.repository.owner.login` in other?

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:278]: I propose to also explicitly specify sort field ("created" is the [default](https://docs.github.com/en/free-pro-team@latest/rest/issues/issues?apiVersion=2022-11-28#list-repository-issues)), as having only direction specified looks a bit weird to my taste.

```suggestion
    sort: "created",
    direction: "asc"
```

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:31]: What happens if a new issue is created during paging? Will the results be distorted, meaning one of the issues will be duplicated in the results? Is there a way to work around this?

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:119]: This is nitpicking, but hey names are important ;) The use of the `github` name is a bit frivolous. You rename the `github.rest` to `octokit` in other places - perhaps stick to that?

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:326]: Why not use the [async](https://nodejs.org/docs/latest-v18.x/api/fs.html#fspromisesreaddirpath-options) versions of the fs functions?

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:345]: Why only check that "metadata.yaml" is part of the file name, as opposed to check that the file name equals "metadata.yaml"?
```suggestion
      } else if (path.basename(filename) == filter) {
```

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [inline .github/workflows/scripts/generate-weekly-report.js:356]: If I understand correctly, by `component` you mean `component class` - a receiver, or a processor, etc. Reading the component type from the path should work, but a cleaner way would be to use the `status.class` field from the metadata file contents.

**Review by andrzej-stencel — APPROVED** [review]
All my comments are minor things that can be either ignored or improved upon in separate PRs. This is good to be merged in my view :+1:

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:262]: when we need to look back for old issues, we're using the hardcoded repo because that's where the actual issues are.
when we're creating a new issue, we use the github context to get the current username in order to create an issue in the forked repo (because its likely that the person that forked the repo doesn't have permissions to create an issue in the collector repo, nor should they if they are testing the script)
i've added a comment to explain this in the code

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:31]: this won't happen because the report will only look at issues created since midnight the day before the action is run (to precisely avoid this case). see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/26111/files#diff-6d378c7867e0f7f3eb58ea282fa1d63937c870c9e55825a4600d3f935592cbb5R73

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:119]: updated

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:326]: the logic was easier to follow when sticking with sync. since the number of directories isn't that large, felt like this wouldn't be a bottleneck

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:345]: good point. updated

File: .github/workflows/scripts/generate-weekly-report.js
Comment (kevinslin) [reply .github/workflows/scripts/generate-weekly-report.js:356]: this is because we currently collect `metadata.yaml` from **everything**. this includes internal components that don't have a `status.class` field (eg. processor/resourcedetectionprocessor/internal/aws/ec2/metadata.yaml). though to be fair. added a check for the `status.class` to set component and default to the current method if nothing is found

**Comment by kevinslin** [comment]
finished addressing pr changes. thanks for the review @astencel-sumo

**Comment by kevinslin** [comment]
checking if there are any other blockers for this pr? @atoulme do you have authorization to merge this pr?

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
I am a little concerned that this introduces javascript to our repository.  @open-telemetry/collector-contrib-approvers is everyone willing to take on this language dependency?

**Comment by kevinslin** [comment]
> I am a little concerned that this introduces javascript to our repository. @open-telemetry/collector-contrib-approvers is everyone willing to take on this language dependency?

its a valid concern. i just want to point out that there is already javascript in the repository (embedded inside of yaml). 
see [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/12760ab56d0724b4cec8315c2c6329a76c9176df/.github/workflows/build-and-test.yml#L644) and [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/12760ab56d0724b4cec8315c2c6329a76c9176df/.github/workflows/milestone-add-to-pr.yml#L25)

**Review by fatsheep9146 — APPROVED** [review]

File: .github/workflows/scripts/generate-weekly-report.js
Comment (astencel-sumo) [reply .github/workflows/scripts/generate-weekly-report.js:330]: > the logic was easier to follow when sticking with sync.

Not sure I understand. The flow of the logic is the same with async, you just add the `await` keyword.

```javascript
    files = await fs.readdir(startPath);
```

The `await` keyword is already used elsewhere in the script, e.g.

```javascript
const response = await octokit.issues.listForRepo(queryParams);
```

I thought maybe it's an issue in importing the `fs/promises` version of the `fs` library?

**Comment by astencel-sumo** [comment]
> I am a little concerned that this introduces javascript to our repository. @open-telemetry/collector-contrib-approvers is everyone willing to take on this language dependency?

I've worked with some decently large Node.js apps in my past, I'd be willing to take it.

Unless we'd rather have this report generator written in a different language, that more contributors of this repo are comfortable. What would it be? Perhaps Python?

**Comment by atoulme** [comment]
I'm fine helping with js code.

**Comment by TylerHelmuth** [comment]
Gonna give one more day for approvers/maintainers to speak up, then I'll merge.

**Pull request merged**
